File: Python/3-Behavioral/State/StateMachine.py


class StateMachine:
    def __init__(self, initial_state):
        # Store the initial state
        self._current_state = initial_state
        # ...and ask run it
        self._current_state.run()

    def run(self, act):
        """
        It allows to execute the new state associated to the given action
        :param act: is the proposed action to execute
        :return: None
        """
        # retrieve the nex state associated to the given action
        self._current_state = self._current_state.next(act)
        # ...and run it
        self._current_state.run()

File: CPP/3-Behavioural/State/StateMachine.cpp
#include "StateMachine.h"
#include "StoppedState.h"

StateMachine::StateMachine()
{
    _currentState = new StoppedState();
}

void StateMachine::run()
{
    _currentState->running(this);
}

void StateMachine::stop()
{
    _currentState->stopped(this);
}
